File: themes/lib/targets/snappy_switcher.py
"""Snappy Switcher theme generator."""

from __future__ import annotations

import os
import shutil
import subprocess

from themes.lib.schema import ColorScheme, ThemeState

TARGET_NAME = "snappy_switcher"
ASSEMBLY = "concat"
OUTPUT_PATH = "~/.config/snappy-switcher/config.ini"
BASE_PATH = "~/repos/dotfiles/config/snappy-switcher/base.ini"
RELOAD_CMD = None
COMMENT = "#"

_THEME_FILE_MAP = {
    "catppuccin-frappe": "catppuccin-mocha.ini",
    "catppuccin-latte": "catppuccin-latte.ini",
    "catppuccin-macchiato": "catppuccin-mocha.ini",
    "catppuccin-mocha": "catppuccin-mocha.ini",
    "gruvbox-dark": "gruvbox-dark.ini",
    "nord": "nord.ini",
    "nord-light": "catppuccin-latte.ini",
    "rose-pine": "rose-pine.ini",
    "rose-pine-dawn": "catppuccin-latte.ini",
    "solarized-dark": "snappy-slate.ini",
    "solarized-light": "catppuccin-latte.ini",
    "tokyo-night": "tokyo-night.ini",
    "tokyo-night-light": "catppuccin-latte.ini",
}


def _rgba(hex_color: str, alpha: str = "ff") -> str:
    """'#rrggbb' -> '#rrggbbaa'."""
    return f"{hex_color}{alpha}"


def _theme_name(state: ThemeState, colors: ColorScheme) -> str:
    """Pick a bundled theme filename as the base theme."""
    theme = _THEME_FILE_MAP.get(state.color_scheme)
    if theme is not None:
        return theme
    if colors.variant == "light":
        return "catppuccin-latte.ini"
    return "snappy-slate.ini"


def generate(colors: ColorScheme, state: ThemeState) -> str:
    title_size = max(state.font_size - 1, 1)
    return f"""\
[theme]
name = {_theme_name(state, colors)}
background = {_rgba(colors.bg)}
card_bg = {_rgba(colors.bg1)}
card_selected = {_rgba(colors.bg2)}
text_color = {_rgba(colors.fg)}
subtext_color = {_rgba(colors.fg4)}
border_color = {_rgba(colors.accent)}
bundle_bg = {_rgba(colors.bg1, "cc")}
badge_bg = {_rgba(colors.accent)}
badge_text_color = {_rgba(colors.bg)}
workspace_color = {_rgba(colors.fg)}
border_width = 2
corner_radius = 12

[icons]
theme = {state.icon_theme}
fallback = Adwaita

[font]
family = {state.system_font}
weight = Bold
title_size = {title_size}
"""


def on_apply(colors: ColorScheme, state: ThemeState) -> None:
    """Restart the daemon so it picks up config changes."""
    del colors, state

    if not os.environ.get("WAYLAND_DISPLAY"):
        return

    snappy = shutil.which("snappy-switcher")
    if snappy is None:
        return

    with open(os.devnull, "w") as devnull:
        subprocess.run(
            [snappy, "quit"],
            check=False,
            stdout=devnull,
            stderr=devnull,
        )
        subprocess.Popen(
            [snappy, "--daemon"],
            stdout=devnull,
            stderr=devnull,
            start_new_session=True,
        )

File: desktopctl/src/theme/targets/snappy_switcher.rs
use super::{Assembly, GeneratedContent, TargetMetadata};
use crate::theme::{
    find_command,
    schema::{ColorScheme, ThemeState},
};
use std::{
    env,
    process::{Command, Stdio},
};

pub const METADATA: TargetMetadata = TargetMetadata {
    name: "snappy_switcher",
    assembly: Assembly::Concat,
    output_path: Some("~/.config/snappy-switcher/config.ini"),
    base_path: Some("~/repos/dotfiles/config/snappy-switcher/base.ini"),
    extra_outputs: &[],
    reload_cmd: None,
    comment: Some("#"),
    sync_safe: true,
};

fn rgba(hex_color: &str, alpha: &str) -> String {
    format!("{hex_color}{alpha}")
}

pub fn generate(colors: &ColorScheme, state: &ThemeState) -> crate::Result<GeneratedContent> {
    let title_size = (state.font_size - 1).max(1);
    Ok(GeneratedContent::text(format!(
        concat!(
            "[theme]\n",
            "name = {}\n",
            "background = {}\n",
            "card_bg = {}\n",
            "card_selected = {}\n",
            "text_color = {}\n",
            "subtext_color = {}\n",
            "border_color = {}\n",
            "bundle_bg = {}\n",
            "badge_bg = {}\n",
            "badge_text_color = {}\n",
            "workspace_color = {}\n",
            "border_width = 2\n",
            "corner_radius = 12\n",
            "\n",
            "[icons]\n",
            "theme = {}\n",
            "fallback = Adwaita\n",
            "\n",
            "[font]\n",
            "family = {}\n",
            "weight = Bold\n",
            "title_size = {}\n",
        ),
        colors.snappy_switcher_theme_name(),
        rgba(&colors.bg, "ff"),
        rgba(&colors.bg1, "ff"),
        rgba(&colors.bg2, "ff"),
        rgba(&colors.fg, "ff"),
        rgba(&colors.fg4, "ff"),
        rgba(&colors.accent, "ff"),
        rgba(&colors.bg1, "cc"),
        rgba(&colors.accent, "ff"),
        rgba(&colors.bg, "ff"),
        rgba(&colors.fg, "ff"),
        state.icon_theme,
        state.system_font,
        title_size,
    )))
}

pub fn on_apply(_colors: &ColorScheme, _state: &ThemeState) -> crate::Result<()> {
    if env::var_os("WAYLAND_DISPLAY").is_none() {
        return Ok(());
    }

    let Some(snappy) = find_command("snappy-switcher") else {
        return Ok(());
    };

    let devnull = Stdio::null();
    let _ = Command::new(&snappy)
        .arg("quit")
        .stdout(devnull)
        .stderr(Stdio::null())
        .status();

    let _child = Command::new(snappy)
        .arg("--daemon")
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn()?;

    Ok(())
}
